enterprise/coderd/identityprovider/authorize.go

```diff
@@ -0,0 +1,160 @@
+package identityprovider
+
+import (
+	"database/sql"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/coderd/database/dbtime"
+	"github.com/coder/coder/v2/coderd/httpapi"
+	"github.com/coder/coder/v2/coderd/httpmw"
+	"github.com/coder/coder/v2/codersdk"
+	"github.com/coder/coder/v2/cryptorand"
+)
+
+type authorizeParams struct {
+	clientID     string
+	redirectURL  *url.URL
+	responseType codersdk.OAuth2ProviderResponseType
+	scope        []string
+	state        string
+}
+
+func extractAuthorizeParams(r *http.Request, callbackURL string) (authorizeParams, []codersdk.ValidationError, error) {
+	p := httpapi.NewQueryParamParser()
+	vals := r.URL.Query()
+
+	p.Required("state", "response_type", "client_id")
+
+	// TODO: Can we make this a URL straight out of the database?
+	cb, err := url.Parse(callbackURL)
+	if err != nil {
+		return authorizeParams{}, nil, err
+	}
+	params := authorizeParams{
+		clientID:     p.String(vals, "", "client_id"),
+		redirectURL:  p.URL(vals, cb, "redirect_uri"),
+		responseType: httpapi.ParseCustom(p, vals, "", "response_type", httpapi.ParseEnum[codersdk.OAuth2ProviderResponseType]),
+		scope:        p.Strings(vals, []string{}, "scope"),
+		state:        p.String(vals, "", "state"),
+	}
+
+	// We add "redirected" when coming from the authorize page.
+	_ = p.String(vals, "", "redirected")
+
+	if err := validateRedirectURL(cb, params.redirectURL.String()); err != nil {
+		p.Errors = append(p.Errors, codersdk.ValidationError{
+			Field:  "redirect_uri",
+			Detail: fmt.Sprintf("Query param %q is invalid", "redirect_uri"),
+		})
+	}
+
+	p.ErrorExcessParams(vals)
+	return params, p.Errors, nil
```

Comment:
Makes sense!  That does seem better.

Reply:
Oh right I needed to check the length from the caller to tell if it was the `url.Parse(callback)` that failed (although that should not happen because we validate before storing it in the db) or if it was a validation error so I can give a different message and error code.  I think I will just lift the parse out of this func.

Reply:
You can also return a sentinel error. So either a unique type or a unique message.

Reply:
Ohh right that is smart!